specs/projectreuniondeployment/projectreuniondeployment.md

```diff
@@ -0,0 +1,138 @@
+present, or may not be at the desired version. This API enables a developer to detect when the
+required Project Reunion main and/or framework packages are present, and to direct the user to the
+appropriate resources to acquire or update Project Reunion if it is absent or requires updates.
+
+Since this API is part of the Project Reunion framework, the framework itself must be present.
+For MSIX packages, the framework must be listed as a dependency in the application manifest and
+will not be deployable without a min version being present. For Win32 programs and installers the
+Project Reunion bootstrapper will try to load the framework and error out with a similar
+instruction on how to resolve the missing dependencies Project Reunion.
+
+Developers must not assume that the Project Reunion main package is present, or that the version
+present is the minimum version required to support the feature set they expect. Instead, this API
+should be used to validate any Project Reunion requirements and direct the user to resolve any
+missing dependencies.
+
+The developer scenarios for this API is to answer the questions:
+1. Does this system have the min version of Project Reunion I require?
+2. What do I do to resolve this if it doesn't?
+
+The API addresses #1 by a simple query of the version they seek and a possible check for OS
+updates. #2 has two possible answers. First is direct resolution by the developer by acquiring and
+installing the packages if possible. Some programs will not have permission or motivation to do
+this, so the second answer is providing a link up to the user where the user can take action to
+install the package. The API will address this scenario by providing an answer as to what is
+required, where to get it, and where to go for instruction.
+
+# Examples
+
+## GetStatus
+
+This is the means by which an app or program can call to verify Project Reunion is present to a
+minimum version. It should be called once after activation before trying to use Project Reunion
+functionality. This method returns a ProjectReunionStatus object that describes the current
+deployment state of Project Reunion.
+
+```C++
+Windows::ApplicationModel::PackageVersion version{1, 0, 0, 0};
+auto status =
+Microsoft::Management::Deployment::ProjectReunionDeployment::GetStatus(version);
+if (!status.IsOK())
+{
+    ...
+}
+```
+
+## GetInstallPackageLink
+
+Provides a direct install link to deploy the latest version of Project Reunion. If no OS Updates
+are required and the program or app has permissions to deploy packages, this can directly resolve
+the missing dependency without burdening the user.
+
+```C++
+Windows::Foundation::Uri packageLink = Microsoft::Management::Deployment::ProjectReunionDeployment::GetInstallPackageLink();
+```
+
+## GetInstallInformationLink
+
+Provides a user-friendly link to acquire the latest version of Project Reunion and any OS updates
```

Comment:
Popping up a link to go install some it of software seems ok for developers (although i'm personally always annoyed and frazzled when this happens in VS). It does not seem like something a non-technical user of Windows should ever experience

it would be good to know what scenarios this can happen in so that developers don't shoot themselves in the foot by going down this route

Reply:
I agree. Each developer is going to be exposing this information differently. I'd rather this is an OS or installer feature, and giving users a consistent experience and gives a lot of confidence and trust in the process of installing more stuff.

Reply:
> the experience of informing the user should be a standard experience

Standard how? By new code in Windows, future, present _and past_? By ProjectReunion providing a 'standard UI'...which devs can (and often will, and should) ignore in favor of better integrating into their own user experience?

This seems like a good simple / poor-man's solution. For the 'more integrated' experience @stevenbrix and @dotMorten advocate why wouldn't you use GetInstallPackageLink() to programmatically handle it as you choose, through your own user experience? (Where 'you' = the app itself, the app's installer, an app's troubleshooter tool, etc)

Reply:
I generally agree with everything here, but there isn't a great solution to this problem. We have downlevel OS updates to deliver, and for packaged apps they will likely not have the capabilities to install other packages. The ideal solution is for the developer to provide the experience in their installer, if possible. If not possible (because they're packaged, for example), then there isn't much else they can do other than direct the user to where they can resolve the problem. For full-trust win32 programs, they could install much of it, and possibly even kick off KBs for the OS updates, but that stuff can change and will be different depending on platform, so its best for us to provide a better experience there. I didn't want to re-write Windows Update here, or re-write their applicability logic for determining which updates are going to be required, that's just going to lead to more problems. 

We have an escape hatch for any developer - link to install project reunion. For more privileged programs and packages and where OS updates are not required, we can get them an appinstaller link so they can kick off the update of the package themselves so the user doesn't have to get involved. But that's really above and beyond expectations for developers. The low barrier to entry and adoption for developers is to redirect to a link we control that can guide them through the process. This should be a 1-time deal, where the first Project Reunion consuming program or package does this, while all subsequent ones benefit from it having already been done before them so they have nothing to do.